File: app/src/main/java/dev/yiray/qickmvivm/ui/home/TaskInteractor.java
package dev.yiray.qickmvivm.ui.home;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.yiray.qickmvivm.model.Task;

public class TaskInteractor {
    public HomeViewState updateTaskList(HomeViewState.ViewState prevStatus, HomeViewState.ListModified modified) {
        List<Task> currentTasks = prevStatus.getTaskList();

        int position = modified.getPosition();
        Task task = modified.getTask();

        if (currentTasks == null) {
            return new HomeViewState.ViewState.Builder()
                    .setTaskList(Collections.singletonList(task))
                    .build();
        }

        ArrayList<Task> current = new ArrayList<>(currentTasks);

        if (current.size() >= position + 1) {
            current.set(position, task);
        } else {
            current.add(task);
        }

        return new HomeViewState.ViewState.Builder()
                .setTaskList(Collections.unmodifiableList(current))
                .build();
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/ui/home/TaskInteractor.kt
package dev.yiray.livedatamvi.ui.home

import dev.yiray.livedatamvi.ui.home.HomeViewState.ListModified
import java.util.*

class TaskInteractor {
    fun updateTaskList(
        prevStatus: HomeViewState.ViewState?,
        modified: ListModified
    ): HomeViewState {
        val currentTasks = prevStatus?.taskList
        val position = modified.position
        val task = modified.task
        if (currentTasks == null) {
            return HomeViewState.ViewState(
                taskList = listOf(task),
                lastModifiedPosition = position
            )
        }
        val current = ArrayList(currentTasks)
        if (current.size >= position + 1) {
            current[position] = task
        } else {
            current.add(task)
        }
        return HomeViewState.ViewState(
            taskList = Collections.unmodifiableList(current),
            lastModifiedPosition = position
        )
    }
}